### DISCUSSION TITLE
Uninstall Previous Per-User Install In Per-Machine Scope?

### DISCUSSION THREAD

**Discussion by NoMoreFood** [wixtoolset/issues#8794] [Questions] [answered]
When running interactively, I want my installer to remove a previously-installed per-user exe-based application but it is not executing as expected.  MSIEXEC returns error 1721 in a verbose log indicating the package is bad.  Any ideas how to accomplish this or what I am doing wrong?

https://github.com/windirstat/windirstat/blob/master/setup/WinDirStat.wxs

**Comment by robmen** [comment]
Would need to see a verbose log file.

**Reply by NoMoreFood** [reply]
Here is it.

[Output.log](https://github.com/user-attachments/files/17438877/Output.log)

**Comment by robmen** [comment, marked as answer]
It turns out the log file wasn't necessary, but seeing it helped illuminate the problem. Your `CustomAction` element uses the `Property` attribute.

```xml
<CustomAction Id="DoUninstallHKCU" Property="UninstallHKCU" ExeCommand="[WIN_DIR_STAT_HKCU] /s"
              Execute="immediate" Impersonate="no" Return="ignore"/>
```

Per the documentation for `CustomAdction Property` it specifies the executable to be executed:

> This attribute is also used with the ExeCommand attribute to create a type 50 custom action that uses the value of the given property to specify the path to the executable.

That means the value of the Property `UninstallHKCU` is used as the executable to launch. However, there is no `UninstallHKCU` Property in your code. So the custom action is executing the `""` program and passing the arguments "path to the uninstaller" and `/s` to it.

The fix should be easy.

```xml
<CustomAction Id="DoUninstallHKCU" Property="WIN_DIR_STAT_HKCU" ExeCommand="/s"
              Execute="immediate" Impersonate="no" Return="ignore"/>
```

Now, it should execute the searched "path to the uninstaller" and pass the `/s` argument to it.

PS: I don't often answer questions here with this much detail (too busy with customers), but I've used WinDirStat with great success in the past. Thank you for your contributions to it.

PPS: The above is a good example of [support offered to FireGiant customers](https://www.firegiant.com/services/).

**Reply by NoMoreFood** [reply]
@robmen Thank you much for the advice and for your WinDirStat patronage!  What you provided worked for when I updated the registry value to not be double quoted which is much further than how far I gotten.  Since I didn't know how to make it strip the quotes, I saw that I can use Directory attribute instead with ExeCommand so I updated the [code](https://github.com/windirstat/windirstat/blob/master/setup/WinDirStat.wxs) and it seems work.  If you don't mind, could you validate I'm not doing completely inadvisable?

Update: Now I just need to figure out how to make it work consistently when the MSI is launched from a non-privileged context.

Thanks!

**Reply by NoMoreFood** [reply]
I ended up pushing execution through CMD which seemed to address the UAC conflict that was coming into play.

**Reply by robmen** [reply]
I can see using `Directory` instead to work around the lack of Windows Installer support for quoted paths in `Property` (I forget about that limitation). I guess I can kinda' see how the `cmd` workaround would work, but I'd use [WixShellExec](https://wixtoolset.org/docs/tools/wixext/util/#wixshellexec-custom-actions) since that'd be cleaner all the way around.

**Reply by NoMoreFood** [reply]
Thank you; I will look into that.
